module/zfs/txg.c

```diff
@@ -549,8 +549,16 @@ txg_sync_thread(void *arg)
 			txg_thread_wait(tx, &cpr, &tx->tx_quiesce_done_cv, 0);
 		}
 
-		if (tx->tx_exiting)
-			txg_thread_exit(tx, &cpr, &tx->tx_sync_thread);
+		/*
+		 * Wait until the quiesce thread has exited to ensure every
+		 * quiesced txg has been synced before exiting.
+		 */
+		if (tx->tx_exiting) {
+			while (tx->tx_threads != 1)
+				txg_thread_wait(tx, &cpr, &tx->tx_exit_cv, 0);
+			if (tx->tx_quiesced_txg == 0)
```

Comment:
shouldn't this be `if (tx->tx_threads == 1)` or just `txg_thread_exit()`? I don't see how `tx_quiesced_txg` could be zero here

Reply:
I am not sure I understand the concern. 

In general, a wait for a cv is almost always enclosed in a loop. 

As to why not `txg_thread_exit()` directly, you said:

> I don't see how tx_quiesced_txg could be zero here

If `tx_quiesced_txg` is nonzero, then there is still a txg quiesced but not synced. I guess it's not a good idea to `txg_thread_exit()` abruptly without syncing it?

As to why  `tx_quiesced_txg` could be zero, if a txg has just been synced, then `txg_quiesce_thread` could wake up first after `txg_sync_stop` broadcasts `tx_quiesce_more_cv`. It sees `tx->tx_exiting` is true so exits immediately without quiescing more txg. Then when `txg_sync_thread` runs, `tx_quiesced_txg` is zero.

Or maybe you meant to say "could be **nonzero** here"?

In that case I hope I have already explained my concern in Motivation and Context above.

Reply:
Would it be sufficient to just change this to:

```
if (tx->tx_exiting && tx->tx_threads == 1) {
        txg_thread_exit(tx, &cpr, &tx->tx_sync_thread);
}
```